library/math/arith.lisp

```diff
@@ -446,199 +398,115 @@ The function general/ is partial, and will error produce a run-time error if the
-    (log  (:a -> :a -> :a)))
-
-  (declare sqrt ((Into Fraction :a) (Exponentiable :a) => :a -> :a))
-  (define (sqrt x)
-    "Returns the square root of the input"
-    (expt x (into 1/2)))
-
-  (define-class ((Into Fraction :a) (Reciprocable :a) (Trigonometric :a) (Exponentiable :a)
-                 => Float :a)
-    "A finite, floating-point approximation of a real or imaginary number"
-    (ee :a)
-    (pi :a))
-
-  (declare exp ((Float :f) => :f -> :f))
-  (define exp
-    "Raise ee to the power of X"
-    (expt ee))
-
-  (declare ln ((Float :f) => :f -> :f))
-  (define ln
-    "The natural logarithm of a given value"
-    (log ee))
-
-  ;; See http://clhs.lisp.se/Body/f_sinh_.htm
-
-  (declare sinh ((Float :f) => :f -> :f))
-  (define (sinh x)
-    (/ (- (exp x) (exp (negate x))) 2))
-
-  (declare cosh ((Float :f) => :f -> :f))
-  (define (cosh x)
-    (/ (+ (exp x) (exp (negate x))) 2))
-
-  (declare tanh ((Float :f) => :f -> :f))
-  (define (tanh x)
-    (/ (sinh x) (cosh x)))
-
-  (declare asinh ((Float :f) => :f -> :f))
-  (define (asinh x)
-    (ln (+ x (sqrt (+ 1 (expt x 2))))))
-
-  (declare acosh ((Float :f) => :f -> :f))
-  (define (acosh x)
-    (* 2 (ln (+ (sqrt (/ (+ x 1) 2)) (sqrt (/ (- x 1) 2))))))
-
-  (declare atanh ((Float :f) => :f -> :f))
-  (define (atanh x)
-    (/ (- (ln (+ 1 x)) (ln (- 1 x))) (fromInt 2)))
-
-  (declare atan2 ((Ord :f) (Float :f) => :f -> :f -> :f))
-  (define (atan2 y x)
-    "Computes the two-argument arctangent of y and x, which is roughly the same as (atan (/ y x)) when defined and accounting for the quadrant of the point (x, y)."
-    (match (Tuple (<=> x 0) (<=> y 0))
-      ((Tuple (GT) _)    (atan (/ y x)))
-      ((Tuple (LT) (LT)) (- (atan (/ y x)) pi))
-      ((Tuple (LT) _)    (+ (atan (/ y x)) pi))
-      ((Tuple (EQ) (GT)) (/ pi (fromint 2)))
-      ((Tuple (EQ) (LT)) (/ pi (fromint -2)))
-      ((Tuple (EQ) (EQ)) 0))))
-
-
-(cl:defmacro %define-real-float-instance (coalton-type underlying-type)
-  "Defines the float instances for a lisp floating-point type"
+  (define-instance (Transfinite Single-Float)
+    (define infinity
+      (lisp Single-Float ()
+        float-features:single-float-positive-infinity))
+    (define nan
+      (lisp Single-Float ()
+        float-features:single-float-nan))
+    (define (nan? x)
+      (Lisp Boolean (x)
+        (float-features:float-NaN-p x)))
+    (define (infinite? x)
+      (Lisp Boolean (x)
+        (float-features:float-infinity-p x))))
+
+  (define-instance (Transfinite Double-Float)
+    (define infinity
+      (lisp Double-Float ()
+        float-features:double-float-positive-infinity))
+    (define nan
+      (lisp Double-Float ()
+        float-features:double-float-nan))
+    (define (nan? x)
+      (Lisp Boolean (x)
+        (float-features:float-NaN-p x)))
+    (define (infinite? x)
+      (Lisp Boolean (x)
+        (float-features:float-infinity-p x))))
+
+  (define-instance (Into Integer (Optional Single-Float))
```

Comment:
do we want `TryInto` instances (too)?

Reply:
These should be `TryInto` instead of `Into`.